src/beamlime/applications/handlers.py

```diff
@@ -0,0 +1,173 @@
+
+    @property
+    def duration(self) -> float:
+        return self._stop - self._start
+
+    def register_handlers(self) -> None:
+        self.messenger.register_handler(StopWatch.Stop, self.stop)
+        self.messenger.register_handler(StopWatch.Start, self.start)
+
+    def start(self, _: BeamlimeMessage) -> None:
+        import time
+
+        self._start = time.time()
+        self.debug("StopWatch logging started.")
+
+    def stop(self, _: BeamlimeMessage) -> None:
+        import time
+
+        self._stop = time.time()
+        self.debug("StopWatch logging finished. Total duration: %s", self.duration)
+
+
+class DataReductionHandler(BaseHandler):
+    """Data reduction handler to process the raw data and update the histogram.
+
+    It receives a list of events, and reduces them into a histogram.
+    It also triggers the update of a plot stream node.
+    """
+
+    input_pipe: List[Events]
+    pipeline: WorkflowPipeline
+
+    def __init__(self, messenger: MessageRouter) -> None:
+        self.output_da: Histogrammed
+        self.stream_node: pp.Node
+        super().__init__(messenger)
+
+    def register_handlers(self) -> None:
+        self.messenger.register_awaitable_handler(RawDataSent, self.process_message)
+
+    def format_received(self, data: Any) -> str:
+        return f"{len(data)} pieces of {Events.__name__}"
+
+    def process_first_input(self, da: Events) -> None:
+        first_pulse_time = da[0].coords['event_time_zero'][0]
+        self.pipeline[FirstPulseTime] = first_pulse_time
+
+    def process_first_output(self, data: Histogrammed) -> None:
+        self.output_da = Histogrammed(sc.zeros_like(data))
+        self.info("First data as a seed of histogram: %s", self.output_da)
+        self.stream_node = pp.Node(self.output_da)
+        self.figure = pp.figure1d(
+            self.stream_node,
+            title="Wavelength Histogram",
+            grid=True,
+        )
+
+    def process_data(self, data: Events) -> Histogrammed:
+        self.info("Received, %s", self.format_received(data))
+        self.pipeline[Events] = data
+        return self.pipeline.compute(Histogrammed)
+
+    async def process_message(self, message: RawDataSent | BeamlimeMessage) -> None:
+        if not isinstance(message, RawDataSent):
+            raise TypeError(f"Message type should be {RawDataSent.__name__}.")
+
+        data = message.content.pop(0)
+        if not hasattr(self, "output_da"):
+            self.process_first_input(data)
+            output = self.process_data(data)
+            self.process_first_output(output)
+        else:
+            output = self.process_data(data)
+
+        self.output_da.values += output.values
+        self.stream_node.notify_children("update")
+        await self.messenger.send_message_async(
+            HistogramUpdated(
+                sender=DataReductionHandler,
+                receiver=Any,
+                content=self.figure,
+            )
+        )
```

Comment:
To be honest, I do not get much of the entire daemon/handler thing. Daemons send messages. Handlers send messages. Daemons contain daemons. Daemons contain handlers, but don't do anything with them. Handlers send and process messages that are not their concern. Etc., ...

I cannot tell why the (seemingly?) high complexity and (seemingly high?) coupling between the various components is needed. I thought I roughly understood how things work when I reviewed the prototypes (being "moved" here), but now I am honestly lost again.

I am not sure how to proceed. Proceed and then trim down in subsequent steps until only essentials are left (and refactor to remove coupling)?

Reply:
Sorry for confusing... 
I renamed this class to ``Application``.
So the idea was to have both ``daemons`` and ``handlers`` under the ``Application``.
``daemon``  should have a ``run`` method, that can be tasks in the event loop and
``handler`` should have a method that can be called for handling certain type of event/message.